libraries/AP_LQR_Control/AP_LQR_Control.cpp

```diff
@@ -0,0 +1,409 @@
+
+    // Calculate the NE position of the aircraft relative to WP A
+    Vector2f A_air = location_diff(prev_WP, _current_loc);
+
+    // calculate distance to target track, for reporting
+    _crosstrack_error = A_air % AB;
+
+    _crosstrack_error=- _crosstrack_error;
+    
+    //set minimum turn radius equal to twice the ground speed
+    float min_turn_rad = 2*groundSpeed;
+    
+    //limit maximum lateral acceleration and set default if zero
+    if((_max_latacc > (100*groundSpeed*groundSpeed/min_turn_rad)) || (_max_latacc ==0))
+    { 
+        _max_latacc = 100*groundSpeed*groundSpeed/min_turn_rad;
+    }
+    
+    //Caluclate adaptive gains
+    float q1= sqrtf(float((_max_xtrack*0.01)/(fabs((_max_xtrack*0.01)-_crosstrack_error))));
+    
+    //Calculate the approach velocity towards path
+    float si = RadiansToCentiDegrees(get_yaw_rad())*0.01;
+    float si_p = (get_bearing_cd(prev_WP,next_WP))*0.01;
+    float temp_sin=sinf(radians(si - si_p));
+    float v_d= groundSpeed * temp_sin;
+    
+    //Compute lateral acceleration based on current state and adaptive gains
+    float u =  - (((_xtrack_fac*0.01)*q1*_crosstrack_error)+(sqrtf((float)((_q2_val*0.01)+(2*q1)))*(_vel_fac*0.01)*v_d));
+    
+    //Limit calculated lateral acceleration to maximum values
+    if (fabs(u) > (_max_latacc*0.01))
+    {
+        if(u > 0)
+            u = _max_latacc*0.01;
+        if(u < 0)
+            u = - (_max_latacc*0.01);
+    }
+    _latAccDem= u;
+
+    
+    // Waypoint capture status is always false during waypoint following
+    _WPcircle = false;
+
+    
+
+    _data_is_stale = false; // status are correctly updated with current waypoint data
+}
+
+// update L1 control for loitering
+void AP_LQR_Control::update_loiter(const struct Location &center_WP, float radius, int8_t loiter_direction)
+{
+    struct Location _current_loc;
+    Vector2f _groundspeed_vector = _ahrs.groundspeed_vector();
+
+    // scale loiter radius with square of EAS2TAS to allow us to stay
+    // stable at high altitude
+    radius = loiter_radius(radius);
+    float groundSpeed=_groundspeed_vector.length();
+    Vector2f location_difference=location_diff(center_WP, _current_loc);
+    
+    //Calculate rate change of heading of path
+    float si_p_dot=groundSpeed/location_difference.length();
+    _crosstrack_error = location_difference.length() - radius;
+    
+    //Set minimum turn radius equal to twice the groundspeed
+    float min_turn_rad = 2*groundSpeed;
+    
+    //limit maximum lateral acceleration and set default if zero
+    if((_max_latacc > (100*groundSpeed*groundSpeed/min_turn_rad)) || (_max_latacc ==0))
+    { 
+        _max_latacc = 100*groundSpeed*groundSpeed/min_turn_rad;
+    }
+    
+    float u =0;
+    
+    // check if vehicle is not very far from the desired circular path
+    if (_crosstrack_error < (radius))
+    {
+        //Compute desired heading perpendicular
+        float si_p = ((_target_bearing_cd + 9000)*0.01);
```

Comment:
How _target_bearing_cd is calculated for loiter mode? It should be calculated as a perpendicular to a vector from a WP (center of loiter) to current position.

Reply:
_target_bearing_cd is being calculated as the bearing towards the center. However, the bearing of path used for actual computation of lateral acceleration is calculated by adding 90 degrees.

Reply:
My point was that in this function _target_bearing_cd is not calculated at all

Reply:
Corrected. I must have missed that line while correcting some stuff.